src/toxcore/dht/server.rs

```diff
@@ -424,37 +462,42 @@ impl Server {
             Ok(payload) => payload,
         };
 
-        if payload.id == 0 {
-            return Box::new( future::err(
-                Error::new(ErrorKind::Other,
-                    "NodesResponse.ping_id == 0"
-            )))
-        }
+        let mut ping_checker = self.ping_checker.write();
+        let ping_checker = ping_checker.deref_mut();
 
-        let client = state.peers_cache.get(&packet.pk);
-        let client = match client {
-            None => {
+        let timeout_dur = Duration::from_secs(PING_TIMEOUT);
+        let nodes_req_data_used = match ping_checker.is_valid(payload.id, timeout_dur) {
+            Err(_err) => {
                 return Box::new( future::err(
                     Error::new(ErrorKind::Other,
-                        "get_client() failed in handle_nodes_resp()"
+                        "NodesResponse::received packet's ping_id does not match or timed out."
                 )))
             },
-            Some(client) => client,
+            Ok(data) => data,
         };
 
-        if client.ping_id == payload.id {
+        if nodes_req_data_used.len() >= 1 {
+            if nodes_req_data_used[0].saddr != addr || nodes_req_data_used[0].pk != packet.pk {
+                return Box::new( future::err(
+                    Error::new(ErrorKind::Other,
+                        "NodesResponse::received packet's address or PK don't match()."
+                )))
+            }
+
+            // TODO: replace it with addto_list
             for node in &payload.nodes {
                 // not worried about removing evicted nodes from peers_cache
                 // they will be removed by timeout eventually since we won't
                 // ping them anymore
                 state.kbucket.try_add(node);
             }
-            Box::new( future::ok(()) )
-        } else {
-            Box::new( future::err(
-                Error::new(ErrorKind::Other, "NodesResponse.ping_id does not match")
-            ))
         }
+
+        if nodes_req_data_used.len() == 2 {
```

Comment:
Try to use custom struct here instead of vector with these magic comparisons.

Reply:
If you need to store different types of data in ping checker you can make it generic.